zip-0317.html

```diff
@@ -260,75 +260,76 @@
                 <p>Miners, mining pools, and other block producers, select transactions for inclusion in blocks using a variety of criteria. The algorithm in the following section is planned to be implemented by <cite>zcashd</cite> and <cite>zebrad</cite>.</p>
                 <section id="recommended-algorithm-for-block-template-construction"><h4><span class="section-heading">Recommended algorithm for block template construction</span><span class="section-anchor"> <a rel="bookmark" href="#recommended-algorithm-for-block-template-construction"><img width="24" height="24" class="section-anchor" src="assets/images/section-anchor.png" alt=""></a></span></h4>
-                    <p>Define a constant
-                        <span class="math">\(weight\_cap = 4\)</span>
+                    <p>Define constants
+                        <span class="math">\(weight\_ratio\_cap = 4\)</span>
+                     and
+                        <span class="math">\(block\_unpaid\_action\_limit = 50\!\)</span>
                     .</p>
                     <p>Let
                         <span class="math">\(conventional\_fee(tx)\)</span>
                      be the conventional fee for transaction
                         <span class="math">\(tx\)</span>
                      calculated according to the section <a href="#fee-calculation">Fee calculation</a>.</p>
+                    <p>Let
+                        <span class="math">\(unpaid\_actions(tx) = \mathsf{max}\!\left(0,\, \mathsf{max}(grace\_actions,\, tx.\!logical\_actions) - \mathsf{floor}\!\left(\frac{tx.fee}{marginal\_fee}\right)\right)\!\)</span>
+                    .</p>
+                    <p>Let
+                        <span class="math">\(block\_unpaid\_actions(block) = \sum_{tx \,\in\, block}\, unpaid\_actions(tx)\)</span>
+                    .</p>
                     <p>The following algorithm is RECOMMENDED for constructing block templates from a set of transactions in a node's mempool:</p>
                     <ol type="1">
                         <li>For each transaction
                             <span class="math">\(tx\)</span>
                          in the mempool, calculate
-                            <span class="math">\(tx.\!weight = \mathsf{min}\!\left(\frac{tx.fee}{conventional\_fee(tx)}, weight\_cap\right)\!\)</span>
-                        .</li>
-                        <li>Repeat while there is any mempool transaction that pays at least the conventional fee, is within the block sigop limit <a id="footnote-reference-7" class="footnote_reference" href="#sigop-limit">4</a>, and fits in the block:
+                            <span class="math">\(tx.\!weight\_ratio = \mathsf{min}\!\left(\frac{\mathsf{max}(1,\, tx.fee)}{conventional\_fee(tx)},\, weight\_ratio\_cap\right)\!\)</span>
```

Comment:
This is 1 zatoshi, not 1 ZEC. It has the effect of making the corner case where all remaining transactions pay zero fee, well defined. (It would otherwise involve computing 0/0 using the most obvious implementation of picking a random transaction by weight.)